### PR TITLE
Native compliance phase

### PR DESCRIPTION
**Pull request by lamont-granquist**

### CODE REVIEW

**Comment by tas50** [comment]
Fat finger here. Sorry

File: lib/chef/resource/inspec_waiver.rb [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+require_relative "../resource"
+
+class Chef
+  class Resource
+    class InspecWaiver < Chef::Resource
+      provides :inspec_waiver
+      unified_mode true
+
+      description "Use the **inspec_waiver** resource to add a waiver to the Compliance Phase."
+      introduced "17.4"
```
Comment (tas50) [inline lib/chef/resource/inspec_waiver.rb:26]: These are all 17.5 now

File: lib/chef/resources.rb [hunk 2]
```diff
@@ -73,6 +73,8 @@
 require_relative "resource/homebrew_tap"
 require_relative "resource/homebrew_update"
 require_relative "resource/ifconfig"
+require_relative "resource/inspec_input"
+require_relative "resource/inspec_waiver"
 require_relative "resource/inspec_waiver_file_entry"
```
Comment (tas50) [inline lib/chef/resources.rb:78]: Is this enough different to keep it around or should we deprecate this resource?

Comment (lamont-granquist) [reply lib/chef/resources.rb:78]: should probably get deprecated, but i don't entirely understand what it is for

File: kitchen-tests/cookbooks/end_to_end/attributes/default.rb [hunk 3]
```diff
@@ -71,3 +71,6 @@
 #
 
 default["nscd"]["server_user"] = "nobody" unless platform_family?("suse") # this breaks SLES 15
+
+# enable CLI output for the compliance phase
+default["audit"]["reporter"] = "cli"
```
Comment (tas50) [inline kitchen-tests/cookbooks/end_to_end/attributes/default.rb:76]: Isn't this the default as of Infra Client 17.4?

**Comment by tas50** [comment]
Paging @chef/docs-team for review on the documentation for these new resources

File: spec/unit/resource/inspec_input_spec.rb [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+    it "enables the input with a regular expression for the cookbook" do
+      load_input("default.yml")
+      resource.name "acme_comp.*"
+      resource.run_action(:add)
+      expect(run_context.input_collection.first).to be_enabled
+      expect(resource).not_to be_updated_by_last_action
+    end
+
+    it "enables the input with an explicit name" do
```
Comment (marcparadise) [inline spec/unit/resource/inspec_input_spec.rb:81]: I don't think there's a test here for success when cookbook is given and it should load default., name 'acme_compliance' with a default present.

**Review by marcparadise — APPROVED** [review]
This looks solid. Some minor comments left inline. None are blockers.

File: lib/chef/compliance/input.rb [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+      #
+      def enable!
+        events.compliance_input_enabled(self)
+        @enabled = true
+      end
+
+      # Set the input as being disabled
+      #
+      def disable!
+        @enabled = false
```
[the remark this answers is not in the document]
Comment (lamont-granquist) [reply lib/chef/compliance/input.rb:73]: I kind of doubt it?  That would get called from the initializer as well which is a bit useless, and I think nothing else calls that (at least not yet).

File: lib/chef/compliance/input_collection.rb [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+      end
+
+      # Add a input to the input collection.  The cookbook_name needs to be determined by the
+      # caller and is used in the `include_input` API to match on.  The path should be the complete
+      # path on the host of the yml file, including the filename.
+      #
+      # @param path [String]
+      # @param cookbook_name [String]
+      #
+      def from_file(filename, cookbook_name)
```
[the remark this answers is not in the document]
Comment (lamont-granquist) [reply lib/chef/compliance/input_collection.rb:40]: yeah, design choice was to disable by default, but then it is very easy to enable everything with e.g. `include_profile ".*::.*"` in any recipe file which is parsed.

File: lib/chef/run_context/cookbook_compiler.rb [hunk 7]
```diff
@@ -134,6 +164,30 @@ def compile_ohai_plugins
         @events.ohai_plugin_load_complete
       end
 
+      # Loads the compliance segment files from the cookbook into the collections
+      # hanging off of the run_context, for later use in the compliance phase
+      # inspec run.
+      #
+      def compile_compliance
+        @events.compliance_load_start
```
[the remark this answers is not in the document]
Comment (lamont-granquist) [reply lib/chef/run_context/cookbook_compiler.rb]: done

File: lib/chef/compliance/input_collection.rb [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+      #
+      # @example Matching inputs by regexp in any cookbook in the cookbook collection
+      #
+      # include_input ".*::ssh.*"
+      #
+      def include_input(arg)
+        raise "include_input was given a nil value" if arg.nil?
+
+        # if we're given a hash argument just shove it in the raw_hash
+        if arg.is_a?(Hash)
```
[the remark this answers is not in the document]
Comment (lamont-granquist) [reply lib/chef/compliance/input_collection.rb:95]: done

File: lib/chef/compliance/input.rb [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+      end
+
+      # Render the input in a way that it can be consumed by inspec
+      #
+      def for_inspec
```
[the remark this answers is not in the document]
Comment (lamont-granquist) [reply lib/chef/compliance/input.rb]: done

**Comment by tas50** [comment]
If this passes I vote merge. If we have additional tests here let's get an issue cut and merge those post 17.5

**Pull request merged**